sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/csharp/CSharpDtoGenerator.java

```diff
@@ -0,0 +1,1355 @@
+
+            generateRecordPropertyAssignment(
+                sb,
+                DefinitionKind.Message,
+                groupToken,
+                indent,
+                groupDtoClassName + ".DecodeListFrom(codec." + formattedPropertyName + ")",
+                "new List<" + groupDtoClassName + ">(0).AsReadOnly()",
+                null
+            );
+
+            i++;
+            i += tokens.get(i).componentTokenCount();
+
+            final List<Token> fields = new ArrayList<>();
+            i = collectFields(tokens, i, fields);
+
+            final List<Token> groups = new ArrayList<>();
+            i = collectGroups(tokens, i, groups);
+
+            final List<Token> varData = new ArrayList<>();
+            i = collectVarData(tokens, i, varData);
+        }
+    }
+
+    private void generateVarDataDecodeFrom(
+        final StringBuilder sb,
+        final List<Token> tokens,
+        final String indent)
+    {
+        for (int i = 0; i < tokens.size(); i++)
+        {
+            final Token token = tokens.get(i);
+            if (token.signal() == Signal.BEGIN_VAR_DATA)
+            {
+                final String propertyName = token.name();
+                final Token varDataToken = Generators.findFirst("varData", tokens, i);
+                final String characterEncoding = varDataToken.encoding().characterEncoding();
+
+                final String formattedPropertyName = formatPropertyName(propertyName);
+                final String accessor = characterEncoding == null ?
+                    "Get" + formattedPropertyName + "Bytes" :
+                    "Get" + formattedPropertyName;
+
+                generateRecordPropertyAssignment(
+                    sb,
+                    DefinitionKind.Message,
+                    token,
+                    indent,
+                    "codec." + accessor + "()",
+                    "null",
+                    null
+                );
+            }
+        }
+    }
+
+    private void generateRecordPropertyAssignment(
+        final StringBuilder sb,
+        final DefinitionKind rootKind,
+        final Token token,
+        final String indent,
+        final String presentExpression,
+        final String notPresentExpression,
+        final String nullCodecValueOrNull)
+    {
+        final String propertyName = token.name();
+        final String formattedPropertyName = formatPropertyName(propertyName);
+
+        sb.append(indent).append(formattedPropertyName).append(": ");
+
+        // N.B., in the IR, the composite information is "embedded" into each message/group field.
+        // When we are looking at a composite "field" (i.e., a type) the `token.version()` method
+        // returns the "sinceVersion" of the containing field. Therefore, we cannot decide presence
+        // based on `token.version()` when dealing with composites.
+        if (rootKind.equals(DefinitionKind.Message) && token.version() > 0)
```

Comment:
I think `token.version() > 0` isn't correct. We should compare `token.version()` against the "parent version".  Where "parent version" might be the `sinceVersion` on the outer composite, message, or group.

N.B., the `BEGIN_GROUP` IR token doesn't hold the version in which the group was added.

Also, see other instances of this pattern.